core/src/main/java/jenkins/model/Jenkins.java

```diff
@@ -3035,6 +3040,36 @@ private void loadConfig() throws IOException {
             // load from disk
             cfg.unmarshal(Jenkins.this);
         }
+
+        setBuildsAndWorkspacesDir();
+    }
+
+    private void setBuildsAndWorkspacesDir() throws IOException {
+        boolean mustSave = false;
+        String newBuildsDir = SystemProperties.getString(BUILDS_DIR_PROP);
+        if (newBuildsDir != null && !buildsDir.equals(newBuildsDir)) {
+            LOGGER.log(Level.WARNING, "Changing builds directories from {0} to {1}. Beware that no automated data migration will occur.",
+                       new String[]{buildsDir, newBuildsDir});
+            buildsDir = newBuildsDir;
+            mustSave = true;
+        } else if (!isDefaultBuildDir()) {
+            LOGGER.log(Level.INFO, "Using non default builds directories: {0}.", buildsDir);
+        }
+
+        String newWorkspacesDir = SystemProperties.getString(WORKSPACES_DIR_PROP);
+        if (newWorkspacesDir != null && !workspaceDir.equals(newWorkspacesDir)) {
+            LOGGER.log(Level.WARNING, "Changing workspaces directories from {0} to {1}. Beware that no automated data migration will occur.",
+                       new String[]{workspaceDir, newWorkspacesDir});
+            workspaceDir = newWorkspacesDir;
+            mustSave = true;
+        } else if (!isDefaultWorkspaceDir()) {
+            LOGGER.log(Level.INFO, "Using non default workspaces directories: {0}.", workspaceDir);
+        }
+
+        if (mustSave) {
+            save();
```

Comment:
seems scary as this is being called in a thread that is performing a load

Reply:
Recommended by @jglick in https://github.com/jenkinsci/jenkins/pull/3364#discussion_r176423228

I'm not totally sure myself of the criticality of the risk, if there is one.